be/src/agent/task_worker_pool.cpp

```diff
@@ -692,6 +693,46 @@ void* ReportWorkgroupTaskWorkerPool::_worker_thread_callback(void* arg_this) {
     return nullptr;
 }
 
+void* ReportResourceUsageTaskWorkerPool::_worker_thread_callback(void* arg_this) {
+    auto* worker_pool_this = (ReportResourceUsageTaskWorkerPool*)arg_this;
+
+    TReportRequest request;
+    AgentStatus status = STARROCKS_SUCCESS;
+
+    while ((!worker_pool_this->_stopped)) {
+        auto master_address = get_master_address();
+        if (master_address.port == 0) {
+            // port == 0 means not received heartbeat yet
+            // sleep a short time and try again
+            sleep(config::sleep_one_second);
+            continue;
+        }
+
+        StarRocksMetrics::instance()->report_resource_usage_requests_total.increment(1);
+        request.__set_backend(BackendOptions::get_localBackend());
+        request.__set_report_version(g_report_version.load(std::memory_order_relaxed));
+
+        TResourceUsage resource_usage;
+        resource_usage.__set_num_running_queries(ExecEnv::GetInstance()->query_context_mgr()->size());
+        resource_usage.__set_mem_used_bytes(ExecEnv::GetInstance()->process_mem_tracker()->consumption());
```

Comment:
Shall we distinguish `query` `cache` `storage` and various memory consumption ?

Reply:
Maybe not, various types of memory are our internal concept, and users may not know how to set thresholds.